Mistral Optimization & Benchmarking Support (#18225)
microsoft/onnxruntime: onnxruntime/python/tools/transformers/models/llama/README.md

# Contents
 - [LLaMA-2](#llama-2)

>    - [Exporting LLaMa-2](#exporting-llama-2)

Reviewer: ```suggestion
   - [Exporting LLaMA-2](#exporting-llama-2)
```

>    - [Benchmarking LLaMa-2](#benchmark-llama-2)

Reviewer: ```suggestion
   - [Benchmarking LLaMA-2](#benchmark-llama-2)
```

    --num-runs 1000 \
    --timeout 60  # number of minutes before moving to the next benchmark
```

> # Mistral

Reviewer: Can we add a table of contents at the beginning to show there's sections for both LLaMA and Mistral?

Author: Good idea! Included in latest commit/

## Introduction

> These tools for LLAMA also allow the quantization and optimizaton of Mistral in ORT. 

Reviewer: ```suggestion
These tools for LLaMA-2 also allow the quantization and optimization of Mistral in ORT. 
```

## Exporting Mistral

There is currently one supported way to export Mistral to ONNX format:

> 

Reviewer: Can Mistral be exported with the `convert_to_onnx` tool?

Author: The last time I tried, I hit some issues with the torchscript exporter and had to fall back to optimum. We could incorporate the optimum export into the `convert_to_onnx` tool, but this PR needs to be merged soon for the AI dev conference so it might be best suited as another PR.

### [Hugging Face Optimum](https://github.com/huggingface/optimum)

The following command will export Mistral in full precision:
```
python -m optimum.exporters.onnx -m mistralai/Mistral-7B-v0.1 --library-name transformers /path/to/model/directory
```

## Optimizing & Quantizing Mistral

To quantize Mistral to FP16 and apply fusion optimizations, you can run the following command:
```

> python models/llama/convert_to_onnx.py -i /path/to/model/directory -o /path/to/optimized_model/directory -p fp16 --optimize_optimum -m mistralai/Mistral-7B-v0.1

Reviewer: ```suggestion
python -m models.llama.convert_to_onnx -i /path/to/model/directory -o /path/to/optimized_model/directory -p fp16 --optimize_optimum -m mistralai/Mistral-7B-v0.1
```

```

## Benchmark Mistral
The benchmarking script in the llama directory (`models/llama/benchmark.py`) directory support Mistral benchmarking by including the flags `-bt mistral-ort` for benchmarking the ONNX model produced by the above, or `-bt mistral-hf`, which will benchmark the HuggingFace implementation of Mistral. For example, to benchmark the ORT and HF versions respectively, you can run: 

```

> python models/llama/benchmark.py -bt ort-convert-to-onnx -p fp16 -m mistralai/Mistral-7B-v0.1 --ort-model-path /dev_data/petermca/mistral/model_opt/model.quant.onnx

Reviewer: ```suggestion
python -m models.llama.benchmark \
    -bt ort-convert-to-onnx \
    -p fp16 \
    -m mistralai/Mistral-7B-v0.1 \
    --ort-model-path /path/to/model.onnx
```

> python models/llama/benchmark.py -bt hf-pt-eager -p fp16 -m mistralai/Mistral-7B-v0.1

Reviewer: ```suggestion
python -m models.llama.benchmark \
    -bt hf-pt-eager \
    -p fp16 \
    -m mistralai/Mistral-7B-v0.1
```

    --num-runs 1000 \
    --timeout 60  # number of minutes before moving to the next benchmark
```

# Mistral

## Introduction

These tools for LLaMA-2 also allow the quantization and optimization of Mistral in ORT. 

## Exporting Mistral

There is currently one supported way to export Mistral to ONNX format:

### [Hugging Face Optimum](https://github.com/huggingface/optimum)

The following command will export Mistral in full precision:
```
python -m optimum.exporters.onnx -m mistralai/Mistral-7B-v0.1 --library-name transformers /path/to/model/directory
```

## Optimizing and Quantizing Mistral

To quantize Mistral to FP16 and apply fusion optimizations, you can run the following command:
```
python -m models.llama.convert_to_onnx -i /path/to/model/directory -o /path/to/optimized_model/directory -p fp16 --optimize_optimum -m mistralai/Mistral-7B-v0.1
```

## Benchmark Mistral
The benchmarking scripts in the LLaMA directory support Mistral benchmarking. To benchmark the ORT and HF versions respectively, you can run: 

```
python -m models.llama.benchmark \
    -bt ort-convert-to-onnx \
    -p fp16 \
    -m mistralai/Mistral-7B-v0.1 \
    --ort-model-path /path/to/model.onnx

> python -m models.llama.benchmark \

Reviewer: Can we separate the two commands into their own code blocks? It can help clarify which command will benchmark which version.

    --num-runs 1000 \
    --timeout 60  # number of minutes before moving to the next benchmark
```

# Mistral

## Introduction

These tools for LLaMA-2 also allow the quantization and optimization of Mistral in ORT. 

## Exporting Mistral

There is currently one supported way to export Mistral to ONNX format:

### [Hugging Face Optimum](https://github.com/huggingface/optimum)

The following command will export Mistral in full precision:
```
python -m optimum.exporters.onnx -m mistralai/Mistral-7B-v0.1 --library-name transformers /path/to/model/directory
```

## Optimizing and Quantizing Mistral

To quantize Mistral to FP16 and apply fusion optimizations, you can run the following command:
```
python -m models.llama.convert_to_onnx -i /path/to/model/directory -o /path/to/optimized_model/directory -p fp16 --optimize_optimum -m mistralai/Mistral-7B-v0.1
```

## Benchmark Mistral
The benchmarking scripts in the LLaMA directory support Mistral benchmarking. To benchmark the ORT version respectively, you can run: 

```
python -m models.llama.benchmark \
    -bt ort-convert-to-onnx \
    -p fp16 \
    -m mistralai/Mistral-7B-v0.1 \
    --ort-model-path /path/to/model.onnx
```

> And to benchmark the Huggingface implementation:

Reviewer: ```suggestion
And to benchmark the Hugging Face implementation:
```
